File: example/src/main/scala/com/github/j5ik2o/pekko/persistence/effector/example/scalaimpl/defaultStyle/nonPersistence/fp/BankAccountAggregateState.scala
package com.github.j5ik2o.pekko.persistence.effector.example.scalaimpl.defaultStyle.nonPersistence.fp

import com.github.j5ik2o.pekko.persistence.effector.example.scalaimpl.{
  BankAccountEvent,
  BankAccountId,
}

enum BankAccountAggregateState {
  def aggregateId: BankAccountId

  case NotCreated(aggregateId: BankAccountId)
  case Created(aggregateId: BankAccountId, bankAccount: BankAccount)

  def applyEvent(event: BankAccountEvent): BankAccountAggregateState = (this, event) match {
    case (
          BankAccountAggregateState.NotCreated(aggregateId),
          BankAccountEvent.Created(id, limit, balance, _)) =>
      Created(id, BankAccount.create(id, limit, balance).bankAccount)
    case (
          BankAccountAggregateState.Created(id, bankAccount),
          BankAccountEvent.CashDeposited(_, amount, _)) =>
      bankAccount
        .add(amount)
        .fold(
          error => throw new IllegalStateException(s"Failed to apply event: $error"),
          result => BankAccountAggregateState.Created(id, result._1),
        )
    case (
          BankAccountAggregateState.Created(id, bankAccount),
          BankAccountEvent.CashWithdrew(_, amount, _)) =>
      bankAccount
        .subtract(amount)
        .fold(
          error => throw new IllegalStateException(s"Failed to apply event: $error"),
          result => BankAccountAggregateState.Created(id, result._1),
        )
    case _ =>
      throw new IllegalStateException(
        s"Invalid state transition: $this -> $event",
      )
  }
}

File: example/src/main/kotlin/com/github/j5ik2o/pekko/persistence/effector/example/kotlinimpl/standard/BankAccountAggregateState.kt
package com.github.j5ik2o.pekko.persistence.effector.example.kotlinimpl.standard

import com.github.j5ik2o.pekko.persistence.effector.example.scalaimpl.BankAccountEvent
import com.github.j5ik2o.pekko.persistence.effector.example.scalaimpl.BankAccountId

/**
 * BankAccountAggregateStateクラス - Kotlin実装
 */
sealed class BankAccountAggregateState {
    abstract val aggregateId: BankAccountId

    /**
     * 未作成状態
     */
    data class NotCreated(override val aggregateId: BankAccountId) : BankAccountAggregateState()

    /**
     * 作成済み状態
     */
    data class Created(
        override val aggregateId: BankAccountId,
        val bankAccount: BankAccount
    ) : BankAccountAggregateState()

    /**
     * イベントを適用して状態を更新する
     */
    fun applyEvent(event: BankAccountEvent): BankAccountAggregateState {
        return when (val state = this) {
            is NotCreated -> {
                when (event) {
                    is BankAccountEvent.Created -> {
                        Created(
                            event.aggregateId(),
                            BankAccount.create(event.aggregateId(), event.limit(), event.balance())
                        )
                    }
                    else -> throw IllegalStateException("Invalid state transition: $state -> $event")
                }
            }
            is Created -> {
                when (event) {
                    is BankAccountEvent.CashDeposited -> {
                        state.bankAccount.add(event.amount()).fold(
                            { error -> throw IllegalStateException("Failed to apply event: $error") },
                            { result -> Created(state.aggregateId, result) }
                        )
                    }
                    is BankAccountEvent.CashWithdrew -> {
                        state.bankAccount.subtract(event.amount()).fold(
                            { error -> throw IllegalStateException("Failed to apply event: $error") },
                            { result -> Created(state.aggregateId, result) }
                        )
                    }
                    else -> throw IllegalStateException("Invalid state transition: $state -> $event")
                }
            }
        }
    }
}
